packages/rrweb-snapshot/src/snapshot.ts

```diff
@@ -279,46 +279,114 @@ export function classMatchesRegex(
   node: Node | null,
   regex: RegExp,
   checkAncestors: boolean,
-): boolean {
-  if (!node) return false;
+  distance: number = -1,
+): number {
+  if (!node) return -1;
   if (node.nodeType !== node.ELEMENT_NODE) {
-    if (!checkAncestors) return false;
+    if (!checkAncestors) return -1;
     return classMatchesRegex(node.parentNode, regex, checkAncestors);
   }
 
   for (let eIndex = (node as HTMLElement).classList.length; eIndex--; ) {
     const className = (node as HTMLElement).classList[eIndex];
     if (regex.test(className)) {
-      return true;
+      return Math.max(distance, 0);
     }
   }
-  if (!checkAncestors) return false;
-  return classMatchesRegex(node.parentNode, regex, checkAncestors);
+  if (!checkAncestors) return -1;
+  return classMatchesRegex(
+    node.parentNode,
+    regex,
+    checkAncestors,
+    distance + 1,
+  );
+}
+
+export function closestMatch(el: HTMLElement, selector: string): number {
+  if (!el) return -1;
+  if (el.matches(selector)) return 0;
+  const closestParent = el.closest(selector);
+  if (closestParent) {
+    let current = el;
+    let distance = 0;
+    while (current && current !== closestParent) {
+      current = current.parentNode as HTMLElement;
+      if (!current) {
+        return -1;
+      }
+      distance++;
+    }
+    return distance;
+  }
+  return -1;
 }
 
 export function needMaskingText(
   node: Node,
   maskTextClass: string | RegExp,
   maskTextSelector: string | null,
+  unmaskTextClass: string | RegExp,
+  unmaskTextSelector: string | null,
+  maskAllText: boolean,
 ): boolean {
   const el: HTMLElement | null =
     node.nodeType === node.ELEMENT_NODE
       ? (node as HTMLElement)
       : node.parentElement;
   if (el === null) return false;
 
-  if (typeof maskTextClass === 'string') {
-    if (el.classList.contains(maskTextClass)) return true;
-    if (el.closest(`.${maskTextClass}`)) return true;
-  } else {
-    if (classMatchesRegex(el, maskTextClass, true)) return true;
+  let mask = !!maskAllText;
+  let maskDistance = -1;
+  let unmaskDistance = -1;
+
+  if (mask) {
```

Comment:
need to clean up a little duplication between mask/unmask, but for discussion:

currently unmask does not work independently of the maskAll option due to the order of operations. it _could_ work but would be less performant. We'd have to always run the unmasking checks, regardless of the current masking status. The current code only runs the unmasking checks if we've already decided to mask, and that can only happen at this point if mask all is true.

There's a similar ordering issue with the class and selector options. If class matches first, we don't bother checking selector, even if it might be a "closer" match. This can also be resolved, but again, would be even less performant, because we'd be checking class & selector every time for both mask and unmask.

Reply:
latest changes should resolve conversation